pkg/stub/handler.go

```diff
@@ -0,0 +1,162 @@
+package stub
+
+import (
+	"context"
+
+	"github.com/integr8ly/tutorial-web-app-operator/pkg/apis/integreatly/v1alpha1"
+
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+	"github.com/sirupsen/logrus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/integr8ly/tutorial-web-app-operator/pkg/apis/integreatly/openshift"
+	"github.com/openshift/api/template/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"fmt"
+	"github.com/operator-framework/operator-sdk/pkg/util/k8sutil"
+	"src/github.com/pkg/errors"
+	errors2 "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/dynamic"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func NewHandler(m *Metrics, osClient openshift.OSClient, dynamicResourceClientFactory func(apiVersion, kind, namespace string) (dynamic.ResourceInterface, string, error)) sdk.Handler {
+	return &Handler{
+		metrics: m,
+		osClient: osClient,
+		dynamicResourceClientFactory: dynamicResourceClientFactory,
+	}
+}
+
+type Metrics struct {
+	operatorErrors prometheus.Counter
+}
+
+type Handler struct {
+	// Metrics example
+	metrics *Metrics
+	osClient openshift.OSClient
+	dynamicResourceClientFactory func(apiVersion, kind, namespace string) (dynamic.ResourceInterface, string, error)
+}
+
+func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
+	switch o := event.Object.(type) {
+	case *v1alpha1.WebApp:
+		if o.Status.Phase == v1alpha1.NoPhase {
+			logrus.Infof("Starting provisioning phase")
+			exts, err := h.processTemplate(o)
+			if err != nil {
+				logrus.Errorf("%v", err)
+			}
+
+			runtimeObjs, err := h.getRuntimeObjs(exts)
+			if err != nil {
+				logrus.Errorf("%v", err)
+			}
+
+			err = h.provisonObjects(runtimeObjs, o)
+			if err != nil {
+				logrus.Errorf("%v", err)
+			}
+
+			logrus.Info("App provisioning started")
+
+			o.Status.Phase = v1alpha1.ProvisionPhase
+			sdk.Update(o)
+		} else if o.Status.Phase == v1alpha1.ProvisionPhase {
+			if h.isAppReady(o) {
+				logrus.Infof("App is ready")
+				o.Status.Phase = v1alpha1.DonePhase
+				sdk.Update(o)
+				logrus.Infof("Provisioning finished")
+			} else {
+				logrus.Infof("Provisioning app")
+
+			}
+		} else if o.Status.Phase == v1alpha1.DonePhase {} else {
```

Comment:
We should probably continually try to be provisioning the web app and exiting gracefully if everything is already there and present.
``` if err != nil && IsAlreadyExists(err){
    
}
```

Reply:
I see you do this below in the create. But I think we want to be always trying to create the objects

Reply:
I doesn't really returns an error and keeps running.

I added this empty block for the sake the error logs, so it wouldn't be too noisy once the app has been deployed.

Reply:
in this case I don't think we don't really need a cr status/phase but some kind of error checking if that makes sense